docs/proposals/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,481 @@
+
+### Assumptions
+
+- Namespaces are pre-populated with one or more `ServiceAccounts` that define the permissions for each `AppProject`.
+- Many users prefer to control access to kubernetes resources through kubernetes RBAC constructs instead of Argo specific constructs.
+- Each tenant is generally given access to a specific namespace along with a service account, role or cluster role and role binding to control access to that namespace.
+- An `AppProject` can either be mapped to a single tenant or multiple related tenants and the respective destinations that needs to be managed via the `AppProject`, needs to be configured.
+
+
+### Goals
+- Applications may only impersonate ServiceAccounts that live in the same namespace as the application itself. ServiceAccount to be used for syncing each application is determined by the target destination configured in the `AppProject` associated with the `Application`.
+- If impersonation feature is enabled, and no service account name is provided in the Application CR, then it should use a default service account of the `Application` namespace.
+- Access restrictions implemented through properties in AppProject (if done) must have the existing behavior. From a security standpoint, any restrictions that were available before switching to a service account based approach should continue to exist even when the impersonation feature is enabled.
+
+### Non-Goals
+
+What is out of scope for this proposal? Listing non-goals helps to focus discussion and make progress.
+None
+
+## Proposal
+
+Enable ArgoCD Admin to specify a `serviceAccountName` in `AppProjects` CR on a per `destination` basis.
+When applications gets synced, based on its destination, the `serviceAccountName` configured in the `AppProject` will be chosen and used for impersonation when executing the kubectl apply/create commands. There could either be a regex based match of the destination or an exact match of the destination. If there are multiple possible options, then an error would be thrown during the sync operation.
+
+```
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: my-project
+  namespace: argocd
+  finalizers:
+    - resources-finalizer.argocd.argoproj.io
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: guestbook
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-deployer
+  - namespace: guestbook-dev
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-dev-deployer
+  - namespace: guestbook-stage
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-stage-deployer
```

Comment:
Supporting full principal name would be preferable, eg. `system:serviceaccount:guestbook-stage:guestbook-stage-deployer`, eg whatever you would pass on to in the `--as` flag.

This allows easier understanding of what flag it gets translated to.
Also supports use-cases when you want to put the deployer service accounts in a separate namespace, not the same namespace that you're targeting.
Eg. SA is in 'deployer' namespace, but targeting 'guestbook-stage'.

Reply:
We want to restrict users from being able to use service account from a different namespace. This is also the convention followed even for pods where you can just specify the `serviceAccountName` and it will be looked up in the pod's namespace.

Reply:
If you go for the pod serviceAccountName equivalence then approject should only use SA from the same namespace and use the `serviceAccountName` field name.

The proposal however seems to follow semantics that are much closer to role binding semantics, which allows for what I described.

Reply:
Some other problems I'm finding:
* how does the audit trail in the k8s audit logs look like if a single application deployment touching multiple namespaces is switching between multiple SA-s?
* how do you control deploying cluster-scoped resources?

From an audit-trail story IMHO it would be a cleaner story if the approject has a clean audit-identity by a 1:1 mapping of AppProject to ServiceAccount.

Reply:
Created an example so its clear what particular deployment model I'm talking about:
https://gist.github.com/reegnz/e83a5958be0d92698061c318f893120e

You have some helm chart, that deploys into one or more namespaces, plus it also deploys cluster-scoped resources.
From an audit perspective you likely don't want the deployment of this particular application to be split between multiple principals. Also you want it to be able to deploy clusterscoped resources and control which clusterscoped resources the app is allowed to deploy.

Reply:
I guess with this proposal the granularity I'm looking for is likely achieved with a 1:1 mapping of AppProject to App.

However it would still be preferable to not assume the deployer SA is in the same namespace as the application, just as we don't assume the AppProject has to reside in the same namespace.
Ideally deployer SA should be allowed to live next to the AppProject IMHO.

Reply:
One example of mine is an application that sets up namespaces + their quotas, limitranges, etc. Essentially bootstrapping the namespaces themselves. It wouldn't make sense splitting them into multiple applications, and we don't have service accounts available for them before creating the namespaces.

Reply:
Another use-case is, when you provision the really critical parts of your cluster, eg. setting up cluster roles to be used by deployer service accounts. You likely want to treat that as a separate app, because creating clusterroles is a VERY sensitive issue.